src/services/DexService.ts

```diff
@@ -1,56 +1,161 @@
+    console.log('User pays for Dex transaction!');
+    const tokenContract = await Contract.initialize({...aeSdk.getContext(), ...{
+      aci: aex9ACI,
+      address: Constants.ae_weth_address,
+    }});
+
+    let { decodedResult: allowance } = await tokenContract.allowance({from_account: aeSdk.address, for_account: Constants.ae_dex_router_address.replace("ct_", "ak_")}, { callStatic: true });
+
+    if (allowance === undefined) {
+      await tokenContract.create_allowance(Constants.ae_dex_router_address.replace("ct_", "ak_"), (amountWei * 2n).toString());
+    } else {
+      if (allowance < amountWei * 2n) {
+        await tokenContract.change_allowance(Constants.ae_dex_router_address.replace("ct_", "ak_"), (amountWei * 2n - allowance).toString());
+      }
+    }
+  }
+
+  // fees are payed by internal wallet
+  static async changeAllowanceInternalPayingFor(amountWei: bigint): Promise<void> {
+    console.log('Internal wallet pays for Dex transaction!');
     const tokenContract = await aeSdk.initializeContract({
       aci: aex9ACI,
-      address: AE_WETH_ADDRESS,
+      address: Constants.ae_weth_address,
     });
 
-    const calldata = tokenContract._calldata.encode(
-      "FungibleTokenFull",
-      "change_allowance",
-      [
-        AE_DEX_ROUTER_ADDRESS.replace("ct_", "ak_"),
-        (amountWei * 2n).toString(),
-      ],
-    );
+    const tokenContractWallet = await Contract.initialize({...aeSdk.getContext(), ...{
+      aci: aex9ACI,
+      address: Constants.ae_weth_address,
+    }});
 
-    const contractCallTx = await aeSdk.buildTx({
-      tag: Tag.ContractCallTx,
-      callerId: aeSdk.address,
-      contractId: AE_WETH_ADDRESS,
-      amount: 0,
-      gasLimit: 1000000,
-      gasPrice: 1500000000,
-      callData: calldata,
-    });
+    let { decodedResult: allowance } = await tokenContractWallet.allowance({from_account: aeSdk.address, for_account: Constants.ae_dex_router_address.replace("ct_", "ak_")}, { callStatic: true });
 
-    const signedContractCallTx = await aeSdk.signTransaction(contractCallTx, {
-      innerTx: true,
-    });
+    if (allowance === undefined) {
+      const calldata = tokenContract._calldata.encode(
+        "FungibleTokenFull",
+        "create_allowance",
+        [
+          Constants.ae_dex_router_address.replace("ct_", "ak_"),
+          (amountWei * 2n).toString(),
+        ],
+      );
+
+      const contractCallTx = await aeSdk.buildTx({
+        tag: Tag.ContractCallTx,
+        callerId: aeSdk.address,
+        contractId: Constants.ae_weth_address,
+        amount: 0,
+        gasLimit: 1000000,
+        gasPrice: 1500000000,
+        callData: calldata,
+      });
+
+      const signedContractCallTx = await aeSdk.signTransaction(contractCallTx, {
+        innerTx: true,
+      });
 
-    await payForTx(signedContractCallTx);
+      await payForTx(signedContractCallTx);
+    } else {
+      console.log("here 2");
+      if (allowance < amountWei * 2n) {
```

Comment:
```suggestion
First of all we should not change allowance if it is enough. Also I don't think that we should expect allowance for twice more amount than user entered. I don't think that slippage can (should) be twice more.
```suggestion
    } else if (allowance < amountWithSlippage) {
```
```